### PR TITLE
feat: implement `update()` for `EvmHookModule`

### PR DESCRIPTION
**Pull request by paulbalaji**
resolves https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/3577

- enables updates of hooks
- includes functionality for updating IGP hook and gas oracles
- drive-by fix to ism module

TODO:

- [x] tests (non-igp)
- [x] tests (igp)

### CODE REVIEW

File: typescript/sdk/src/ism/EvmIsmModule.hardhat-test.ts
Comment (paulbalaji) [reply typescript/sdk/src/ism/EvmIsmModule.hardhat-test.ts:421]: woops nice catch - i removed a bunch after testing but not these

**Comment by ltyu** [comment]
> ### ⚠️ No Changeset found
> Latest commit: [cebab01](https://github.com/hyperlane-xyz/hyperlane-monorepo/commit/cebab01075fc34492b354ee563349e0c2f192ed0)
> 
> Merging this PR will not cause a version bump for any packages. If these changes should not result in a new version, you're good to go. **If these changes should result in a version bump, you need to add a changeset.**
> 
> This PR includes no changesets
> [Click here to learn what changesets are, and how to add one](https://github.com/changesets/changesets/blob/main/docs/adding-a-changeset.md).
> 
> [Click here if you're a maintainer who wants to add a changeset to this PR](https://github.com/hyperlane-xyz/hyperlane-monorepo/new/pb/3577?filename=.changeset/seven-houses-grin.md&value=---%0A%22%40hyperlane-xyz%2Fsdk%22%3A%20patch%0A---%0A%0Afeat%3A%20implement%20%60update()%60%20for%20%60EvmHookModule%60%0A)

add this :D

File: typescript/sdk/src/hook/EvmHookModule.ts
Comment (ltyu) [inline typescript/sdk/src/hook/EvmHookModule.ts:572]: iiuc, `deploy()` will transfer ownership to the `config.owner`. This is fine if the owner and the subsequent signer of this updateTx are the same. 

But i believe it will be an issue if `config.owner` is set to an address that cannot sign txs, such as SAFE, address(0), etc.

example flow for illustration:
- CLI calls this fn with `signer`
- fn deploys, and then transfers to `config.owner`
- CLI attempts to `sendTransaction(updateTxs)` using `signer`

File: typescript/sdk/src/hook/EvmHookModule.ts
Comment (ltyu) [inline typescript/sdk/src/hook/EvmHookModule.ts:365]: iiuc, deploy() will transfer ownership to the config.owner. This is fine if the owner and the subsequent signer of this updateTx are the same.

But i believe it will be an issue if config.owner is set to an address that cannot sign txs, such as SAFE, address(0), etc.

example flow for illustration:
- CLI calls this fn with signer
- fn deploys, and then transfers to config.owner
- CLI attempts to sendTransaction(updateTxs) using signer

File: typescript/sdk/src/hook/EvmHookModule.ts
Comment (ltyu) [inline typescript/sdk/src/hook/EvmHookModule.ts:208]: In which case will this execute? i thought the owner is always transferred, if needed in the deploy fns?

File: typescript/sdk/src/hook/EvmHookModule.ts
Comment (ltyu) [inline typescript/sdk/src/hook/EvmHookModule.ts:472]: There's no validation in these Config schemas. This might lead to weird unexpected behavior. For example,

If `target` has `remoteDomain` key, it will overwrite the fetched `remoteDomain`. Might be intentional, but it's also not a valid `StorageGasOracleConfigSchema`. I believe its possible to have something like 
```typescript
const target: StorageGasOracleConfigSchema = { 
  gasPrice: "1000", 
  tokenExchangeRate: "1000", 
  blablablabla: 7777 
}
```

Consider using Schema.parse to 1) validate config schemas, and 2) filter out extra parameters

I probably have to do this for my PRs too ;)

**Review by ltyu — COMMENTED** [review]
mostly makes sense. i'll need to have another peek at this

File: typescript/sdk/src/hook/EvmHookModule.ts
Comment (paulbalaji) [inline typescript/sdk/src/hook/EvmHookModule.ts:143]: avoid "->" arrow notation, in favour of explicit sentences/words

File: typescript/sdk/src/hook/EvmHookModule.ts
Comment (paulbalaji) [reply typescript/sdk/src/hook/EvmHookModule.ts:365]: the intent with these new modules is that they should read/create/update to the input given by the caller, to reduce the necessity of "owner" checks and implicit behavioural changes that plagued the current suite of deployers

so any `deploy()` here will ensure the owner of the new contract matches the given config's owner - this should mean that there are no follow-on updates required afterwards

so if the CLI does attempt to do _another_ update afterwards, it is expected for that to go through the multisig, as that is the configuration that was provided

in the current world we already have the case where we if do a new deploy for adding more chains, we have to go through a multisig to set up the new ISM as the default on the mailbox

File: typescript/sdk/src/hook/EvmHookModule.ts
Comment (paulbalaji) [reply typescript/sdk/src/hook/EvmHookModule.ts:208]: this can be super useful down the line with checker/governance actions too

File: typescript/sdk/src/hook/EvmHookModule.ts
Comment (paulbalaji) [reply typescript/sdk/src/hook/EvmHookModule.ts:472]: i'm aware it's not great in ts land with objects so i tried to be mindful about only selecting/using fields i require from an object - but I'll add schema parsing at the top of constructor/deploy/update functions in both ISM + Hook modules as it never hurts to be safe!

**Review by ltyu — COMMENTED** [review]
I approve, but will be comfortable if another dev takes a look. @yorhodes

File: typescript/sdk/src/hook/EvmHookModule.hardhat-test.ts
Comment (nbayindirli) [inline typescript/sdk/src/hook/EvmHookModule.hardhat-test.ts:158]: nit: remove

File: typescript/sdk/src/hook/EvmHookModule.hardhat-test.ts
Comment (nbayindirli) [inline typescript/sdk/src/hook/EvmHookModule.hardhat-test.ts:168]: set `let signer: Signer;` in parent scope and

```suggestion
    [signer, funder] = await hre.ethers.getSigners();
```

**Review by nbayindirli — COMMENTED** [review]
partial, still reviewing

File: typescript/sdk/src/hook/EvmHookModule.ts
Comment (nbayindirli) [inline typescript/sdk/src/hook/EvmHookModule.ts:364]: can we follow this same `updateTxs.push(...(await this.*));` pattern above in the `update()` switch-case? this would also update line 168 to `const updateTxs: AnnotatedEV5Transaction[] = [];`

File: typescript/sdk/src/hook/EvmHookModule.ts
Comment (nbayindirli) [inline typescript/sdk/src/hook/EvmHookModule.ts:148]: thoughts on declaring `let updateTxs: AnnotatedEV5Transaction[];` early on, and returning this instead of `[]`?

nit: remove comment and inline like `if (...) return updateTxs;`

**Review by nbayindirli — APPROVED** [review]
thanks for flushing this out!

most of my comments are re codestyle, and therefore readability/maintainability. i believe at this point in the SDK we should err on using variable & function names to convey purpose, probably resulting in a bunch of `private` helper functions in classes like these

one thing that seems useful for repurpose is capturing all the update TX types somewhere so they can be reused through the SDK, up to you if worth the effort rn

otherwise thanks for being so thorough and functionally looks good, approved

File: typescript/sdk/src/hook/EvmHookModule.ts
Comment (paulbalaji) [reply typescript/sdk/src/hook/EvmHookModule.ts:129]: it's to conform to the abstract HyperlaneModule class, where `TConfig` in the `EvmHookModule` is `HookConfig`, we reuse the TConfig in a few places throughout the abstract class to aid typings
image

File: typescript/sdk/src/hook/EvmHookModule.ts
Comment (paulbalaji) [reply typescript/sdk/src/hook/EvmHookModule.ts:135]: not really, because the error is specifically for the case when targetConfig is just a pure string instead of an object

File: typescript/sdk/src/hook/EvmHookModule.ts
Comment (paulbalaji) [reply typescript/sdk/src/hook/EvmHookModule.ts:140]: this is probably a preference thing but I would much rather read a comment than try to infer from variable names

File: typescript/sdk/src/hook/EvmHookModule.ts
Comment (paulbalaji) [reply typescript/sdk/src/hook/EvmHookModule.ts:159]: this is a good candidate for a helper function

File: typescript/sdk/src/hook/EvmHookModule.ts
Comment (paulbalaji) [reply typescript/sdk/src/hook/EvmHookModule.ts:219]: I've thought about doing this but didn't want to overload this PR, because we can extract this out to a wider transfer ownership helper that we can reuse across all the modules instead of just in here

File: typescript/sdk/src/hook/EvmHookModule.ts
Comment (paulbalaji) [reply typescript/sdk/src/hook/EvmHookModule.ts:148]: I usually prefer to `return []` to make it explicit what should happen

**Comment by paulbalaji** [comment]
> i believe at this point in the SDK we should err on using variable & function names to convey purpose, probably resulting in a bunch of `private` helper functions in classes like these

yeah I can agree with that - however I'm not sure I agree with removing comments? I like to sprinkle them through bits of logic so that when read together it paints a clearer picture of what's happening where. yes, we should have var/fn names that indicate purpose but not to the point you _have_ to read the code. imo comments/naming both can help in different ways 

> one thing that seems useful for repurpose is capturing all the update TX types somewhere so they can be reused through the SDK, up to you if worth the effort rn

100% - again this was one of those things that I wanted to do after merging this in and post-tx submitter merging (shows how long this PR has been cooking lol). There's some common update patterns across the modules and I wanted to get a feel for how the modules get implemented before typing them

File: typescript/sdk/src/hook/EvmHookModule.ts
Comment (paulbalaji) [reply typescript/sdk/src/hook/EvmHookModule.ts:364]: not sure I like the verbosity but ig idm too much

File: typescript/sdk/src/hook/EvmHookModule.ts
Comment (paulbalaji) [reply typescript/sdk/src/hook/EvmHookModule.ts:148]: re inline if statements, i'm not really a fan of them at all but will adopt if it makes its way into the linter

keeping as-is because there's probably a bunch of other places that exist where we have a mix of inline and normal

**Pull request merged**
